certification/internal/authn/keychain_test.go

```diff
@@ -0,0 +1,223 @@
+// Copyright 2018 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package authn
+
+import (
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"reflect"
+	"testing"
+
+	craneauthn "github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+)
+
+var (
+	fresh              = 0
+	testRegistry, _    = name.NewRegistry("test.io", name.WeakValidation)
+	testRepo, _        = name.NewRepository("test.io/my-repo", name.WeakValidation)
+	defaultRegistry, _ = name.NewRegistry(name.DefaultRegistry, name.WeakValidation)
+)
+
+func TestMain(m *testing.M) {
+	// Set $HOME to a temp empty dir, to ensure $HOME/.docker/config.json
+	// isn't unexpectedly found.
+	tmp, err := ioutil.TempDir("", "keychain_test_home")
+	if err != nil {
+		log.Fatal(err)
+	}
+	os.Setenv("HOME", tmp)
+	os.Exit(func() int {
+		defer os.RemoveAll(tmp)
+		return m.Run()
+	}())
+}
```

Comment:
`TestMain` seems odd. I would read this as you're testing the `Main()`, which I don't believe is what's happening here. General convention maps gotest functions to look something like `Test<FunctionUnderTest><Permutation>`. I won't say we should strictly follow that convention, but maybe something other than Main here would be sufficient.

Reply:
Is resetting HOME here necessary? The doc for WithAuthFromKeychain seems to imply that the default keychain gets overridden with either your keychain or anonymous just by invoking this option https://github.com/google/go-containerregistry/blob/main/pkg/crane/options.go#L87

Reply:
Yeah, I can change TestMain to something more meaningful. This file was copied directly from go-containerregistry. As for the $HOME thing, no, it doesn't need to be there, since we aren't using that anyway.

Reply:
After removing the HOME reassignment, TestMain was no longer necessary.